HGCTPGStage1Emulator/test/HGCTPGStage1EmulatorTester.cc

```diff
@@ -0,0 +1,235 @@
+  SAConfigParser cfgReader(jsonConfigFile);
+  stage1Config_ = cfgReader.getCfg();
+
+  rzmin_ = stage1Config_.rozMin();
+  rzmax_ = stage1Config_.rozMax();
+  rz_bin_size_ = (rzmax_ - rzmin_) * margin_ / stage1Config_.rozBins();
+
+  theTCMap_ = cfgReader.getTCmap();
+
+  std::cout << "-- Initializing TC reader" << std::endl;
+  SATCParser tcReader(tcList, theTCMap_);
+  std::cout << "-- Getting TC list" << std::endl;
+  allTCs_ = tcReader.TClist();
+
+}
+
+int HGCTPGEmulatorTester::runAlgo(const HGCalTriggerCellsPerBx& tcPerBx, HGCalTriggerCellsPerBx& sortedTruncatedTCsPerBx) {
+  
+  unsigned int counter = 0;
+  for (const auto TCs : tcPerBx) {
+    // Run the algorithm
+    HGCalTriggerCells sortedTruncatedTCs;
+    unsigned error_code = stage1Algo_.run(TCs, stage1Config_, sortedTruncatedTCs);
+    sortedTruncatedTCsPerBx.push_back(sortedTruncatedTCs);
+  }
+  return 0;
+}
+
+int HGCTPGEmulatorTester::dumpTCs(const HGCalTriggerCellsPerBx& tcPerBx,
+				  const std::string outputFileName) {
+  
+  std::ofstream outputStream;
+  outputStream.open(outputFileName);
+  unsigned int counter = 0;
+  for (const auto TCs: tcPerBx) {
+    outputStream << "Event " << counter << std::endl;
+    for (const auto& tc : TCs ) {
+
+      double rz = tc.rOverZ();
+      rz = (rz < rzmin_ ? rzmin_ : rz);
+      rz = (rz > rzmax_ ? rzmax_ : rz);
+      unsigned rzbin = (rz_bin_size_ > 0. ? unsigned((rz - rzmin_) / rz_bin_size_) : 0);
+      
+      outputStream << "TCID" << tc.index() << " : "
+		   << "roverZ = " << tc.rOverZ() << " "
+		   << "(bin " << rzbin << ") :"
+		   << tc.energy() << std::endl;
+    }
+    counter++;
+  }
+  outputStream.close();
+  return 0;
+}
+
+int HGCTPGEmulatorTester::dumpTCs_fw(const HGCalTriggerCellsPerBx& tcPerBx,
+				     const std::string outputFileName) {  
+
+  /* 
+
+     1 file per (roz,phi) bin (a,b): results_Phi(a)Bin(b).txt
+
+     in file:
+     
+      ---- TCs ---->
+    | XXXXYY XXXXYY XXXXYY ... 
+    | XXXXYY XXXXYY XXXXYY ... 
+    B XXXXYY XXXXYY XXXXYY ... 
+    x XXXXYY XXXXYY XXXXYY ... 
+    | XXXXYY XXXXYY XXXXYY ... 
+    | XXXXYY XXXXYY XXXXYY ... 
+    v ...
+    
+    XXXX: 16 bits address (modId/tcId?)
+    YY: 8 bits TC energy
+
+  */
+  
+  // Create TC vectors per (roz,phi) bin
+  unsigned event=0;
+  std::unordered_map< unsigned, std::vector<std::pair<unsigned, HGCalTriggerCell> > > TCandEvtperBin;
+  for (const auto TCs: tcPerBx) {
+    for (const auto& tc: TCs) { 
+      double rz = tc.rOverZ();
+      rz = (rz < rzmin_ ? rzmin_ : rz);
+      rz = (rz > rzmax_ ? rzmax_ : rz);
+      unsigned rzbin = (rz_bin_size_ > 0. ? unsigned((rz - rzmin_) / rz_bin_size_) : 0);
+      
+      double phi = tc.phi();
+      unsigned phibin = (phi > stage1Config_.phiEdges().at(rzbin));
+
+      unsigned packed_bin = 0;
+      packed_bin |= ((rzbin & mask_roz_) << offset_roz_);
+      packed_bin |= (phibin & mask_phi_);
+      
+      TCandEvtperBin[packed_bin].push_back(std::make_pair(event,tc));
```

Comment:
Better to use `emplace_back` to construct the pair in-place and avoid the copy

Reply:
done